### ISSUE TITLE
[CT-775] [Enhancement] `dbt deps` does not correctly resolve local dependency of local dependency

### ISSUE THREAD

**Issue by skirino** [dbt-labs/dbt-core#5410]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Suppose there are the following [local dependency](https://docs.getdbt.com/docs/building-a-dbt-project/package-management#local-packages) relations: `pkg1` => `pkg2` => `pkg3`.
Then, the directory of `pkg3` is specified by `packages.yml` file of `pkg2`. But, when running `dbt deps` in `pkg1`, `pkg3` is searched from `pkg1`'s directory, not from `pkg2`'s directory.

### Expected Behavior

The directory of `pkg3` should be resolved from `pkg2`'s directory, so that `pkg3` can be found regardless of the location of `pkg1`.

### Steps To Reproduce

I've prepared a minimal repro: https://github.com/skirino/dbt_local_dep_path_issue

1. `git clone https://github.com/skirino/dbt_local_dep_path_issue.git`
2. `cd dbt_local_dep_path_issue/b/c`
3. `dbt deps`

    ```
    [2022-06-27 21:01:17] $ dbt deps                       [~/tmp/dbt_local_dep_path_issue/b/c] 0s
    12:01:20  Running with dbt=1.1.1
    12:01:21  Encountered an error:
    Runtime Error
      no dbt_project.yml found at expected path /Users/skirino/tmp/dbt_local_dep_path_issue/b/common2/dbt_project.yml
    ```

### Environment

```markdown
- OS: macOS 12.4
- Python: 3.7.13
- dbt: 1.1.1
```

### What database are you using dbt with?

snowflake

### Additional Context

This is the same issue as described in https://github.com/dbt-labs/dbt-core/issues/4538#issuecomment-1127034448, but I think it deserves a separate issue.

**Comment by emmyoop** [comment]
I always love a good reproduction repo. ❤️   Thanks for that @skirino!

I agree this is related to #4538 but I think it may have to do with `base_path` as well so a separate issue is appropriate.  This will take some more research and I'm not sure of the consequences of supporting this (circular dependencies, specifically).  

I'm also going to change this to an enhancement, because similar to #5374 this was not the original intent on how this would be used.

**Comment by jaklan** [comment]
Bumping the issue, it makes local packages in the monorepo setup completely unusable. And the above @emmyoop's statement:
> this was not the original intent on how this would be used

is not valid anymore, as the monorepo use-case is mentioned directly in local package docs nowadays:  
https://docs.getdbt.com/docs/build/packages#local-packages

cc: @jtcohen6, because that's absolutely critical issue for our dbt monoproject refactor

**Comment by jaklan** [comment]
@dbeatty10 could you re-open the issue not to create duplicates?

**Comment by jtcohen6** [comment]
@jaklan Thanks for flagging this - it's worth another look given the changes to packages/progress over the past few years.

Right now, the `local` package's resolved path is always relative to the root project from which you're running `dbt deps`:

https://github.com/dbt-labs/dbt-core/blob/71f3519611a6a21c2cb492df2d54e60af040e0f3/core/dbt/deps/local.py#L44-L48

Using OP's [reproduction case](https://github.com/skirino/dbt_local_dep_path_issue/tree/master), `project.project_root` is always `Path('/Users/.../dbt_local_dep_path_issue/b/c')`. We'd need to find a way to instead pass in the absolute path of the package's `packages.yml`, because what we really want there is `Path('/Users/.../dbt_local_dep_path_issue/common1') + Path('../common2')`.

So we'd need a way to get at the location of the `packages.yml` file in which each `local` package specification is actually defined. That `packages.yml` could very likely be located within the downloads directory, if it belongs to a package which was itself installed via git / Hub / tarball method. I think the trickiness here will be a mix of plumbing (the `deps` classes have several layers of nesting) and of thinking through potential edge cases at the intersection of multiple deps methods.

**Comment by jaklan** [comment]
@jtcohen6 thanks for picking up the issue. I fully agree it gets quite tricky when we consider mix of various installation methods, however - maybe it's worth to split it into two phases: a) temporary solution to fix the local-of-local scenario b) the final, robust one handling various combinations.

I can imagine a use-case, where we install a package via `git` and that package has some local dependency (e.g. a common package in its repo) - but then the complexity is getting really serious, we have to re-think such edge-cases like e.g. sparse checkout when using `subdirectory` for packages relying on local dependencies in another directories:

> If the packaged project is instead nested in a subdirectory—perhaps within a much larger mono repo—you can optionally specify the folder path as subdirectory. dbt will attempt a [sparse checkout](https://git-scm.com/docs/git-sparse-checkout) of just the files located within that subdirectory.

and most likely much more other not-so-obvious scenarios.

But having said that, when we talk about the monorepo setup - I believe the most common scenario is the local-of-local one, and solving only that scenario would be enough to unblock most monorepo users and give you time to re-design the current approach in a more comprehensive way.

**Comment by dbeatty10** [comment]
@misteliy I'm going to take a closer look at the local-of-local use-case during the next sprint.

Does the example in https://github.com/skirino/dbt_local_dep_path_issue encapsulate the local-of-local use-case that you're thinking of?

**Comment by jaklan** [comment]
@dbeatty10 yep, based on the README - that's exactly the issue

**Comment by misteliy** [comment]
@dbeatty10 keep us posted - thank you!

**Comment by dbeatty10** [comment]
@jaklan and @misteliy I took a look at this and put up a draft PR: https://github.com/dbt-labs/dbt-core/pull/10600

It worked when I tested it against [https://github.com/skirino/dbt_local_dep_path_issue](https://github.com/skirino/dbt_local_dep_path_issue), but it's failing a couple CI tests (specifically [these two](https://github.com/dbt-labs/dbt-core/blob/3c558062033e8fde33cbf9f923886edc2842d346/tests/functional/partial_parsing/test_partial_parsing.py#L843-L855)). I haven't yet had a chance to dig into the reason why.

**Comment by jaklan** [comment]
@dbeatty10 thank you for taking care of it, highly appreciated!

**Comment by misteliy** [comment]
Just commenting since it seems not resolved yet

**Comment by sdebruyn** [comment]
Also ran into this during development of tsql_utils and dbt-fabric. Once one transitive package has a local relative dependency, it messes up the whole structure
